### PR TITLE
WIP: refactor: added abstraction layer for encoder and graphics device

### PR DESCRIPTION
**Pull request by karasusan**
I refactored c++ code overall.
It will be easier to add encoder and graphics device by this changes.

### Contents
- Abstracted two modules (Encoder, Graphics device)
- Installed [gtest](https://github.com/google/googletest) framework that possible to refactor safely
- Added CI jobs to test native code

### Class design
image

### TODO
Currently not support Linux native test on Yamato yet, I will work on this soon.

### CODE REVIEW

File: Plugin~/WebRTCPluginTest/ContextTest.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        context = std::make_unique<Context>();
+    }
+    void TearDown() override {
+        EncoderFactory::GetInstance().Shutdown();
+        GraphicsDeviceTestBase::TearDown();
+    }
+};
+TEST_P(ContextTest, InitializeAndFinalizeEncoder) {
+    context->InitializeEncoder(m_device);
+    context->FinalizerEncoder();
```
Comment (sindharta) [inline Plugin~/WebRTCPluginTest/ContextTest.cpp]: r in the middle is probably not necessary ? FinalizeEncoder ?

File: Plugin~/WebRTCPluginTest/GraphicsDeviceTestBase.cpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
+﻿#include "pch.h"
+#include "GraphicsDeviceTestBase.h"
+#include "../WebRTCPlugin/GraphicsDevice/GraphicsDevice.h"
+
+using namespace WebRTC;
+
+#if defined(_WIN32)
```
Comment (sindharta) [inline Plugin~/WebRTCPluginTest/GraphicsDeviceTestBase.cpp]: #if defined(SUPPORT_D3D11) is probably better ?

File: Plugin~/WebRTCPluginTest/GraphicsDeviceTestBase.cpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        s_glutInitialized = true;
+        glutCreateWindow("test");
+    }
+    return nullptr;
+}
+#endif
+
+std::tuple<UnityGfxRenderer, void*> GraphicsDeviceTestBase::CreateParameter()
+{
+#if defined(WIN32)
```
Comment (sindharta) [inline Plugin~/WebRTCPluginTest/GraphicsDeviceTestBase.cpp]: this one too.

File: Plugin~/WebRTCPluginTest/ContextTest.cpp (hunk 1 above)
Comment (karasusan) [reply Plugin~/WebRTCPluginTest/ContextTest.cpp]: fixed

File: Plugin~/WebRTCPluginTest/GraphicsDeviceTestBase.cpp (hunk 2 above)
Comment (karasusan) [reply Plugin~/WebRTCPluginTest/GraphicsDeviceTestBase.cpp]: fixed it

File: Plugin~/WebRTCPluginTest/GraphicsDeviceTestBase.cpp (hunk 3 above)
Comment (karasusan) [reply Plugin~/WebRTCPluginTest/GraphicsDeviceTestBase.cpp]: fixed it

File: Plugin~/WebRTCPlugin/Codec/EncoderFactory.h [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    class IGraphicsDevice;
+    class EncoderFactory {
+    public:
+        static EncoderFactory& GetInstance();
+        bool IsInitialized() const;
+        void Init(int width, int height, IGraphicsDevice* device);
+        void Shutdown();
+        IEncoder *GetEncoder() const;
+    private:
+        std::unique_ptr<IEncoder> m_encoder = nullptr;
```
[the remark this answers is not in the document]
Comment (karasusan) [reply Plugin~/WebRTCPlugin/Codec/EncoderFactory.h]: fixed it

**Comment by sindharta** [comment]
LGTM

**Pull request merged**